lib/main/atom/commands/refactorCode.tsx

```diff
@@ -0,0 +1,131 @@
+import {addCommand, Dependencies} from "./registry"
+import {commandForTypeScript, getFilePathPosition, spanToRange} from "../utils"
+import {selectListView} from "../views/simpleSelectionView"
+import * as etch from "etch"
+import {HighlightComponent} from "../views/highlightComponent"
+import * as protocol from "typescript/lib/protocol"
+import * as atom from "atom"
+import {TypescriptServiceClient} from "../../../client/client"
+
+interface RefactorAction {
+  refactorName: string
+  refactorDescription: string
+  actionName: string
+  actionDescription: string
+  inlineable: boolean
+}
+
+addCommand("atom-text-editor", "typescript:refactor-selection", deps => ({
+  description: "Get a list of applicable refactors to selected code",
+  async didDispatch(e) {
+    if (!commandForTypeScript(e)) {
+      return
+    }
+
+    const editor = e.currentTarget.getModel()
+    const location = getFilePathPosition(editor)
+    if (!location) {
+      e.abortKeyBinding()
+      return
+    }
+    const selection = editor.getSelectedBufferRange()
+    if (selection.isEmpty()) {
+      e.abortKeyBinding()
+      return
+    }
+    const client = await deps.getClient(location.file)
+
+    const fileRange: protocol.FileRangeRequestArgs = {
+      file: location.file,
+      startLine: selection.start.row + 1,
+      startOffset: selection.start.column + 1,
+      endLine: selection.end.row + 1,
+      endOffset: selection.end.column + 1,
+    }
+
+    const actions = await getApplicableRefactorsActions(client, fileRange)
+
+    if (actions.length === 0) {
+      // TODO Show a "no applicable refactors here" message
+      e.abortKeyBinding()
+      return
+    }
+
+    const selectedAction = await selectListView({
+      items: actions,
+      itemTemplate: (item, ctx) => {
+        return (
+          <li>
+            <HighlightComponent
+              label={`${item.refactorDescription}: ${item.actionDescription}`}
+              query={ctx.getFilterQuery()}
+            />
+          </li>
+        )
+      },
+      itemFilterKey: "actionDescription",
+    })
+
+    if (selectedAction !== undefined) {
```

Comment:
For the sake of consistency, I would prefer
```ts
if (!selectedAction) return
await applyRefactors(...)
```

Reply:
Since there was a [strict-boolean-expressions](https://github.com/TypeStrong/atom-typescript/commit/f5ec4f162524365684f91bdb342188cc152b98c9) commit, wouldn't be preferred to use
```
if (selected === undefined) ...
```
instead?

Reply:
We have `allow-null-union` and `allow-undefined-union` enabled. So unless the value can be falsy some other way, boolean coercions like this are allowed.

The reason is simple: non-null objects are never falsy, and it's usually obvious from context whether something's an object or a boolean.

I mean, I'm all for clarity, but there's clear code and then there's code that's unnecessarily verbose.

Reply:
Perfect. No problemo.